docs/scaling-hive.md

```diff
@@ -0,0 +1,47 @@
+# Scaling Hive
+
+This document describes the main places users are likely to see scalability issues with Hive.
+
+## Hive uses CRDs!
+
+Most importantly, be aware that Hive uses CRDs to store its state. The amount of data that Hive can store is limited by the cluster's ability to store CRs. In OpenShift, this boils down to "how much data can we stuff into etcd before the cluster itself becomes unstable?" Best to not chance it. We recommend keeping the total number of Hive CRs in a Hive cluster in the thousands. Anything approaching 10,000 is the danger zone. In practice this means that a single Hive cluster can manage about 1000 clusters.
+
+# Horizontal vs. Vertical Scale
+
+With the exception of install pods (used only when clusters are installing), Hive 1.x is not horizontally scalable at the worker level. Most of the work Hive does happens in the hive-controllers pod, which is one single pod on one single worker. This means that when no installs are running, if you have a cluster with 10 workers, 9 of the workers are very bored. Hive clusters are prime candidates for using worker autoscaling. Keep the worker count as low as you can, but allow bursts of concurrent installs to call for temporary workers to spin up.
+
+In AWS, Hive performs best on c instance types. M instances are fine, but c instances are better.
+
+In general, the faster your etcd storage, the faster Hive will perform. c5.4xlarge has twice the EBS bandwidth as c5.2xlarge, and so Hive performs a bit better on c5.4xlarge (although in our opinion, not so much faster that c5.4xlarge is worth the premium over c5.2xlarge). In AWS, we recommend using io1 with 1200 iops for your masters.
+
+If you need to scale out the number of clusters you need to manage, use multiple Hive clusters -- approximately 1 Hive per 1000 clusters managed. Or fewer if you prefer to keep the blast radius smaller.
+
+## Install Pods
+
+Hive 1.x request 800 Mib of memory for each install pod. If you use m5.xlarge workers, you can support about (15 Gib / 800 Mib) install pods per worker -- so about 16. If you need to support more concurrent installs, you can use more workers, and/or workers with more memory. Install pods use barely any CPU.
+
+## Blocking I/O
+
+hive-controllers (where the syncset controllers run) uses blocking i/o. We may move to scale-out or non-blocking i/o in the future, but for now you should be aware that communication (mostly syncsets) to managed clusters can block other things.
```

Comment:
Not referring to client throttling.

@staebler confirmed we use blocking i/o during a stand up.

If your ss controller has 5 threads, and you send 5 GETs to 5 clusters that are Unavailable, you will do no work in the cc controller for 30 seconds each thread waits to timeout.

Contrast with async i/o like in nodejs where 5 threads send 5 requests, they immediately go do something else, and they are called back when the response comes back. This is much more scalable.

Reply:
> you should be aware that communication (mostly syncsets) to managed clusters can block other things

The communication that the *clustersync* controller does with a remote cluster may block the *clustersync* controller from reconciling other clusters. It will not block other controllers from performing work.

Reply:
We may also want to mention that we detect unavailable clusters and we don't attempt to apply syncsets to them. So this is a temporary issue until those clusters are determined to be unavailable.

Reply:
I attempted to clarify this. I tried to take into account everyone's suggestions.